report/latest_map_generated.go

```diff
@@ -42,23 +42,45 @@ func (m StringLatestMap) Size() int {
 	return len(m)
 }
 
-// Merge produces a fresh StringLatestMap containing the keys from both inputs.
+// Merge produces a StringLatestMap containing the keys from both inputs.
 // When both inputs contain the same key, the newer value is used.
+// Tries to return one of its inputs, if that already holds the correct result.
 func (m StringLatestMap) Merge(n StringLatestMap) StringLatestMap {
 	switch {
 	case m == nil:
 		return n
 	case n == nil:
 		return m
 	}
-	l := len(m)
-	if len(n) > l {
-		l = len(n)
+	lenm := len(m)
+	if len(n) > lenm {
+		m, n, lenm = n, m, len(n) //swap so m is always at least as long as n
 	}
-	out := make([]stringLatestEntry, 0, l)
 
 	i, j := 0, 0
-	for i < len(m) {
+loop:
+	for i < lenm {
+		switch {
+		case j >= len(n) || m[i].key < n[j].key:
+			i++
+		case m[i].key == n[j].key:
+			if m[i].Timestamp.Before(n[j].Timestamp) {
```

Comment:
Looks like this makes Merge non-commutative as far as performance is concerned, e.g. when merging two maps m,n with identical keys & values, where all entries in m are older than n, then the operation will copy, whereas it will simply return n if the args are swapped.

One way to avoid that would be to check the timestamp of the first element of equal-length args, and swap the args if necessary.

Reply:
I would tend to ask for profiling evidence that that case is important.

Reply:
The stated purpose of this PR is to optimise the case where Scope
> merges multiple versions of the same node over time that have exactly the same labels, tags, etc., so we detect this and avoid copying.

If all the above is true, but the timestamps differ then the new code still copies. And I am quite sure the timestamps will nearly _always_ differ. I am also quite sure that in the common case the timestamps will differ systematically s.t. all timestamps in one arg are before that in the other. So we should optimise that case. Yes, it would help to check these "quite sure"s against actual data.

It is possible that the Merge benchmark, or indeed the Merge code, ensures that in most cases the Merge is always invoked as new.Merge(old). But that is not something I'd want to rely on.

Reply:
Also, non-commutativity on performance of an otherwise commutative operation is quite distasteful.

Reply:
Aha, I hadn't previously noticed that the probe, which is caching everything, re-generates most `Node` objects from scratch on each report.

I would expect the merge benchmark is merging from oldest to newest, since the filenames begin with a timestamp, Go sorts filenames, and there is no other sort operation. Bears examination.

Reply:
Maybe it's just evidence that allocating memory is so much slower than everything else that a few extra copies doesn't show up.

Reply:
Examination reveals that `Nodes.UnsafeMerge()` is merging left into right, which reverses the sort order from the merge benchmark. I added a `Shuffle()` to the test to defeat this.